eva/experimental/ray/executor/exchange_executor.py

```diff
@@ -60,25 +62,35 @@ def __init__(self, node: ExchangePlan):
         self.ray_conf = node.ray_conf
         super().__init__(node)
 
+    def _find_all_parent_executor_of_exchange(self, curr_exec):
+        # Traverse the query execution tree in a DFS manner. This
+        # function will return the parent executor of ExchangeExecutor,
+        # which will later be inserted a queue reader.
+
+        # Start with checking if current executor has any child.
+        if len(curr_exec.children) > 0:
+            # Check if child executor is ExchangeExecutor. If so, return.
+            # Otherwise, traverse further down.
+            if isinstance(curr_exec.children[0], ExchangeExecutor):
+                yield curr_exec
```

Comment:
I feel this does not work with tree structure. Consider the following:
```mermaid
graph TD;
    R[EXCHANGE] --> A;
    A[JOIN]-->B[EXCHANGE];
    A-->C[EXCHANGE];
    B --> D[SOME UDF1];
    C --> E[SOME UDF2];
```
We never get to the right subtree.

Reply:
I see. I will need to cover this case then. I mainly attempted to fix the following case based on our test cases. 
```mermaid
graph TD;
E1[Exchange]-->P
P[Project]-->N
N[NestedLoopJoin]-->SL
N-->SR
SL[SeqScan]-->E2
E2[Exchange]-->S1[Storage]
SR[SeqScan]-->E3
E3[Exchange]-->S2[Storage]
```

I think we can revisit the Exchange optimizer now to decide when it is actually needed and we can modify the executor based on that.

Reply:
Got this will work. I think what we want to parallel is, for example, the UDF in the seqscan's projection. Parallelizing the Storage is not really needed. It was a "lazy" implementation, so we don't need to support partition in the Storage. But it is not  optimal.

Reply:
I think `Storage` will still benefit from `Exchange(dop=1)` for async IOs. Otherwise, the disk read needs to happen in a blocking fashion.

My main point here is that we probably have two cases to cover for tree-like execution.
1. There exists multiple children and children are not `Exchange`. I think my current change can handle this case.
2. As you pointed out, a tree can have multiple children, which are `Exchange`. Then, we need add some mechanism for `ExchangeExecutor` to handle this case.

Maybe we have some other cases as well?

Reply:
BTW, join does not work with ray with `parallelism > 1` as for now, because two ray actors will not have the complete information. Maybe tree structure is not a priority to worry about now.